infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/ParallelDockerImagesBuilder.java

```diff
@@ -0,0 +1,305 @@
+ */
+package org.eclipse.che.workspace.infrastructure.docker;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static org.eclipse.che.workspace.infrastructure.docker.DockerMachine.LATEST_TAG;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.inject.assistedinject.Assisted;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.function.Supplier;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.core.util.FileCleaner;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalInfrastructureException;
+import org.eclipse.che.infrastructure.docker.client.DockerConnector;
+import org.eclipse.che.infrastructure.docker.client.DockerFileException;
+import org.eclipse.che.infrastructure.docker.client.ProgressMonitor;
+import org.eclipse.che.infrastructure.docker.client.UserSpecificDockerRegistryCredentialsProvider;
+import org.eclipse.che.infrastructure.docker.client.exception.ImageNotFoundException;
+import org.eclipse.che.infrastructure.docker.client.json.Filters;
+import org.eclipse.che.infrastructure.docker.client.params.BuildImageParams;
+import org.eclipse.che.infrastructure.docker.client.params.ListImagesParams;
+import org.eclipse.che.infrastructure.docker.client.params.PullParams;
+import org.eclipse.che.infrastructure.docker.client.params.TagParams;
+import org.eclipse.che.infrastructure.docker.client.parser.DockerImageIdentifier;
+import org.eclipse.che.infrastructure.docker.client.parser.DockerImageIdentifierParser;
+import org.eclipse.che.workspace.infrastructure.docker.exception.SourceNotFoundException;
+import org.eclipse.che.workspace.infrastructure.docker.logs.MachineLoggersFactory;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerContainerConfig;
+import org.slf4j.Logger;
+
+/**
+ * This class allows to make parallel prepare (build or download) of docker images for workspace
+ * being started.
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class ParallelDockerImagesBuilder {
+
+  private static final Logger LOG = getLogger(ParallelDockerImagesBuilder.class);
+  private static final String PARALLEL_PULL_PROPERTY_NAME = "che.infra.docker.max_pull_thread";
```

Comment:
Please remove this constant or move `"che.docker.always_pull_image"` into constant field too.

Reply:
for what ? it is not used twice

Reply:
Sorry, I've missed that PARALLEL_PULL_PROPERTY_NAME is used twice.